gtsfm/frontend/matcher/cacher/matcher_cacher.py

```diff
@@ -0,0 +1,181 @@
+
+logger = logger_utils.get_logger()
+
+CACHE_ROOT_PATH = Path(__file__).resolve().parent.parent.parent.parent.parent / "cache"
+NUM_KEYPOINTS_TO_SAMPLE_FOR_HASH = 10
+
+
+class MatcherCacher(MatcherBase):
+    """Decorator which caches matcher output, keyed on the input."""
+
+    def __init__(self, matcher_obj: MatcherBase) -> None:
+        super().__init__()
+        self._matcher = matcher_obj
+        # TODO: make the obj cache key dependent on the code
+        self._matcher_obj_key = type(self._matcher).__name__
+
+    def __get_cache_path(self, cache_key: str) -> Path:
+        """Gets the cache path from the cache key."""
+        return CACHE_ROOT_PATH / "matcher" / "{}.pbz2".format(cache_key)
+
+    def __generate_cache_key(
+        self,
+        keypoints_i1: Keypoints,
+        keypoints_i2: Keypoints,
+        descriptors_i1: np.ndarray,
+        descriptors_i2: np.ndarray,
+        im_shape_i1: Tuple[int, int],
+        im_shape_i2: Tuple[int, int],
+    ) -> str:
+        """Generates the cache key from the input detections, image shapes, and underlying matcher."""
+        numpy_arrays_to_hash: List[np.ndarray] = []
+
+        # subsample and concatenate keypoints and descriptors
+        numpy_arrays_to_hash.append(keypoints_i1.coordinates[:NUM_KEYPOINTS_TO_SAMPLE_FOR_HASH].flatten())
+        if keypoints_i1.responses is not None:
+            numpy_arrays_to_hash.append(keypoints_i1.responses[:NUM_KEYPOINTS_TO_SAMPLE_FOR_HASH].flatten())
+        if keypoints_i1.scales is not None:
+            numpy_arrays_to_hash.append(keypoints_i1.scales[:NUM_KEYPOINTS_TO_SAMPLE_FOR_HASH].flatten())
+        numpy_arrays_to_hash.append(descriptors_i1[:NUM_KEYPOINTS_TO_SAMPLE_FOR_HASH].flatten())
+        numpy_arrays_to_hash.append(keypoints_i2.coordinates[:NUM_KEYPOINTS_TO_SAMPLE_FOR_HASH].flatten())
+        if keypoints_i2.responses is not None:
+            numpy_arrays_to_hash.append(keypoints_i2.responses[:NUM_KEYPOINTS_TO_SAMPLE_FOR_HASH].flatten())
+        if keypoints_i2.scales is not None:
+            numpy_arrays_to_hash.append(keypoints_i2.scales[:NUM_KEYPOINTS_TO_SAMPLE_FOR_HASH].flatten())
+        numpy_arrays_to_hash.append(descriptors_i2[:NUM_KEYPOINTS_TO_SAMPLE_FOR_HASH].flatten())
+
+        # add the shapes as a numpy array
+        numpy_arrays_to_hash.append(np.array([im_shape_i1[0], im_shape_i1[1], im_shape_i2[0], im_shape_i2[1]]))
+
+        # hash the concatenation of all the numpy arrays
+        input_key = cache_utils.generate_hash_for_numpy_array(np.concatenate(numpy_arrays_to_hash))
+
+        return "{}_{}".format(self._matcher_obj_key, input_key)
+
+    def __load_result_from_cache(
+        self,
+        keypoints_i1: Keypoints,
+        keypoints_i2: Keypoints,
+        descriptors_i1: np.ndarray,
+        descriptors_i2: np.ndarray,
+        im_shape_i1: Tuple[int, int],
+        im_shape_i2: Tuple[int, int],
+    ) -> Optional[np.ndarray]:
+        """Load cached result, if it exists."""
```

Comment:
might be nice to mention in the docstring what the return np.ndarray represents? shape?